aten/src/ATen/test/apply_utils_test.cpp

```diff
@@ -0,0 +1,112 @@
+  auto t = t_.view(-1);
+  for (int64_t i = 0; i < t.numel(); i++) {
+    t[i] = (i + 1) * scalar;
+  }
+}
+
+void test(Type& type, IntList shape, int64_t a=0, int64_t b=1) {
+  auto zero_dim = type.tensor({});
+  zero_dim.fill_(2);
+  zero_dim.exp_();
+  AT_DISPATCH_FLOATING_TYPES(zero_dim.type(), "test0", [&] {
+    ASSERT(zero_dim.data<scalar_t>()[0] == std::exp(2));
+  });
+
+  auto empty_t = type.tensor({0});
+  empty_t.fill_(3);
+  empty_t.exp_();
+
+  auto a0 = type.tensor();
+  auto a1 = type.tensor();
+  auto a2 = type.tensor();
+  auto a3 = type.tensor();
+  auto a4 = CPU(kDouble).tensor();
+
+  std::vector<Tensor> tensors({a0, a1, a2, a3, a4});
+  for (size_t i = 0; i < tensors.size(); i++) {
+    tensors[i].resize_(shape);
+    fill_tensor(i + 1, tensors[i]);
+    if (a >= 0 && b >= 0) {
+      tensors[i].transpose_(a, b);
+    }
+  }
+
+
+  AT_DISPATCH_FLOATING_TYPES(a0.type(), "test1", [&] {
+    CPU_tensor_apply2<scalar_t, scalar_t>(
+        a0, a1, [](scalar_t& y, const scalar_t& x) { y = x * x; });
+    CPU_tensor_apply2<double, scalar_t>(
+        a4, a1, [](double& y, scalar_t x) { y = (double)(x * x); });
+    for (int64_t i = 0; i < a0.numel(); i++) {
+      auto target = a1.data<scalar_t>()[i] * a1.data<scalar_t>()[i];
+      ASSERT(a0.data<scalar_t>()[i] == target);
+      ASSERT(a4.data<double>()[i] == target);
+    }
+  });
+
+  AT_DISPATCH_FLOATING_TYPES(a0.type(), "test2", [&] {
+    CPU_tensor_apply3<scalar_t, scalar_t, scalar_t>(
+        a0, a1, a2, [](scalar_t& y, const scalar_t& x, const scalar_t& z) {
+          y = x * x + z;
+        });
+    CPU_tensor_apply3<double, scalar_t, scalar_t>(
+        a4, a1, a2, [](double& y, const scalar_t& x, const scalar_t& z) {
+          y = (double)(x * x + z);
+        });
+    for (int64_t i = 0; i < a0.numel(); i++) {
+      auto target = a1.data<scalar_t>()[i] * a1.data<scalar_t>()[i];
+      target = target + a2.data<scalar_t>()[i];
+      ASSERT(a0.data<scalar_t>()[i] == target);
+      ASSERT(a4.data<double>()[i] == target);
+    }
+  });
+
+  AT_DISPATCH_FLOATING_TYPES(a0.type(), "test3", [&] {
+    CPU_tensor_apply4<scalar_t, scalar_t, scalar_t, scalar_t>(
+        a0,
+        a1,
+        a2,
+        a3,
+        [](scalar_t& y,
+           const scalar_t& x,
+           const scalar_t& z,
+           const scalar_t& a) { y = x * x + z * a; });
+    CPU_tensor_apply4<double, scalar_t, scalar_t, scalar_t>(
+        a4,
+        a1,
+        a2,
+        a3,
+        [](double& y, const scalar_t& x, const scalar_t& z, const scalar_t& a) {
+          y = (double)(x * x + z * a);
+        });
+    for (int64_t i = 0; i < a0.numel(); i++) {
+      auto target = a1.data<scalar_t>()[i] * a1.data<scalar_t>()[i];
+      target = target + a2.data<scalar_t>()[i] * a3.data<scalar_t>()[i];
+      ASSERT(a0.data<scalar_t>()[i] == target);
+      ASSERT(a4.data<double>()[i] == target);
+    }
+  });
+}
+
+TEST_CASE("apply utils test", "[cpu]") {
```

Comment:
It would be nice to use the Catch2 features that let you create sub tests. Right now it's really hard to tell what are you trying to check in different places. Otherwise some comments would be useful.